keps/83-workload-preemption/README.md

```diff
@@ -0,0 +1,636 @@
+   they take to finish, on average.
+   This implies that all workloads in the cohort have similar characteristics
+   and all the ClusterQueues in the cohort should have the same wait period.
+
+This caveat should be part of the documentation as a best practice for how to
+setup the field.
+
+#### Why no control to opt-out a ClusterQueue from preemption
+
+In a cohort, some ClusterQueue could have high priority Workloads running, so
+it might be desired not to disturb them.
+
+However, this could be achieved by two means:
+- Restricting other ClusterQueues to only have workloads with lower priority,
+  via [QuotaScopes](https://kubernetes.io/docs/concepts/policy/resource-quotas/#quota-scopes)
+- Configuring the ClusterQueue with high priority Workloads to never borrow
+  (through `.quota.max`), while owning a big part or all of the quota for the
+  cohort.
+
+### Risks and Mitigations
+
+#### Workload preemption doesn't imply immediate Pod termination
+
+When Kueue issues a Workload preemption, the workload API integration controller
+is expected to start removing Pods.
+In the case of Kubernetes' batch.v1/Job, the following steps happen:
+1. Kueue's job controller sets the
+[`.spec.suspend`](https://kubernetes.io/docs/concepts/workloads/controllers/job/#suspending-a-job)
+field to true.
+2. The Kubernetes' job controller deletes the Job's Pods.
+3. The Kubelets send SIGTERM signals to the Pod's containers, which can have a
+   graceful termination logic.
+
+This implies the following:
+- Pods of a workload could implement checkpointing as part of their graceful
+  termination.
+- The resources from these Pods are not immediately available and they could
+  be arbitrarily delayed.
+- While Pods are terminating, a ClusterQueue's quota could be oversubscribed.
+
+Oversubscription could be mitigated by watching the running pods but this will
+be left for future work.
+
+#### No clear mechanism to fail or terminate a Job
+
+There are two means to terminate all running Pods of a Job:
+
+1. Delete the Job. The pods will be deleted (gracefully) on cascade.
+
+   This could mean loss of information for the end-user, unless they have a
+   finalizer on the Job. It also violates
+   [`ttlAfterFinish`](https://kubernetes.io/docs/concepts/workloads/controllers/ttlafterfinished/)
+
+1. Just suspend the Job.
+  
+   This option leaves a Job that is not finished, then `ttlAfterFinish`
+   couldn't clean it up.
+
+   Simply adding a `Failed` condition after suspending the Job could leave its
+   Pods running undefinitely if the kubernetes job controller doesn't have a
+   chance to delete all the Pods based on the `.spec.suspend` field.
+
+<UNRESOLVED>
+As a mitigation, we will delete the Job, but only after `ttlAfterFinish` since
+the time of preemption has passed (from the timestamp in the `Finished`
+condition on the workload).
+
+Jobs that don't define `ttlAfterFinish` will be left untouched indefinitely.
```

Comment:
This seems problematic, unclean. I'm not sure we should be adding option Terminate then.

One worry is that Story 2 is written assuming batch user interacting with Workload. Is this normal interaction in Kueue, or something new? I thought the batch user (scheduling jobs) would rather interact with Job, or with custom workload that was integrated with Kueue, but not with Workload itself. If so, could we rewrite that story to user/Job terms (if we keep the story)?

Another type of user story might be a developer integrating a custom workload with Kueue. We should have guidance for them how to intergate Workload.Finished states. But if it's hard to do for us in Job reference implementation, it will probably be also hard for custom workloads?

Another direction would be that maybe this should be a Job feature instead? I.e. how to behave on suspensions, e.g. fail after suspension. Btw. how do new [Pod retry policies](https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md) treat Job suspension? (CC @mimowo)

Reply:
> Another direction would be that maybe this should be a Job feature instead?

I agree. Although I was trying to provide a mitigation while we upstream such behavior (1.28 at the earliest).

But maybe this can be considered a bug fix https://github.com/kubernetes/kubernetes/issues/113142

> how do new Pod retry policies treat Job suspension?

Those pods are ignored from the policy (the same way they are ignored from the status failed/succeeded counters). This is because during suspension, for each pod, we first remove the finalizer and then delete the Pod. Then, we the pod finishes, it disappears.

I will leave this section for now until we get some feedback in the issue I opened upstream.

Reply:
It seems to me this is actually being introduced under my work. There is a new interim FailureTarget job condition which is translated into Failed job condition along with the regular job termination.

Reply:
Ha, that would actually work, if we are ok with 2 things:

- Support in 1.26+ only
- Triggering deletion from a condition, which might be surprising https://github.com/kubernetes/kubernetes/issues/113142#issuecomment-1282591323